### ISSUE TITLE
Limit the number of instances of a template that can run concurrently

### ISSUE THREAD

**Issue by behrmann** [systemd/systemd#28075]
### Component

systemd

### Is your feature request related to a problem? Please describe

I want to batch a number of instances of a service, i.e. I need to run a large number of services that are similar, but could use e.g. different sandboxing for paths that might be writable or should be inaccessible, and of which only a subset should run at any given time, e.g. because of resource constraints.

One instance where this came up for me was during a large scale ID migration so that two university departments agreed to on UIDs/GIDs.

### Describe the solution you'd like

It would be nice to start all instances at once, but have only have a subset run at any given time, but still have all of them run at some point.

The problem could be generalized to only run a given number of services that are grouped in a slice.

### Describe alternatives you've considered

Alternatives:
- start all instances at once and given them lower CPU/IO/... shares (problem: the services might still fight for resources amongst themselves, even thought other services might be unaffected by them)
- do the batching in the software that is being run (problem: harder to sandbox different instances differently)
- batch manually on service start with `Wants=` and `After=` (problem: very manual and brittle)

**Comment by Werkov** [comment]
This is related to #17097 that is accidentally being tackled in #27985.

> only a subset should run at any given time, e.g. because of resource constraints.

What mechanism should be used to pick the lucky ones?

If resources are tight, services would slow down naturally as they have to share them. Or do you observe nonlinear behavior? (I.e. `n` parallel instances finish later than `n` instances sequentially.)

**Comment by behrmann** [comment]
> What mechanism should be used to pick the lucky ones?

In principle I guess one could get away with not guaranteeing any order (in the absence of `Before=` and `After=`), but  I think FIFO submission order would be easiest to reason about, e.g. I start `foo@a.service` , `foo@b.service` , `foo@c.service`  and `foo@d.service` and only two may run at a given time, so `a` and `b` would run first, with `c` and `d` waiting and `c` would start when either `a` or `b` finishes.

> If resources are tight, services would slow down naturally as they have to share them. Or do you observe nonlinear behavior? (I.e. n parallel instances finish later than n instances sequentially.)

So far in those cases I batched things up manually. Usually the workload was somewhat IO bound and it was rather about making sure that some instances finished first and there was "a visible progress" instead of all instances going ahead at roughly the same pace and finishing together.

**Comment by osnyx** [comment]
There is a chance that https://github.com/systemd/systemd/pull/27985 unfortunately does **not** tackle this in its current state. Reading the OP's use case, we want all the started instances to succeed in the end, but just the ones exceeding a concurrency limit to block until there's room for them again.
The PR above so far reads like units exceeding such a limit will end up in a failed state instead of blocking and waiting, though.

Am I missing a way about how such a fail-on-limit behaviour could be turned into the units (re)starting until the limit is not exceeded anymore?

**Comment by osnyx** [comment]
To quickly outline another instance where such a concurrency limit is useful:

In https://github.com/NixOS/nixpkgs/issues/232505, the generation of TLS certificates for individual domains is encapsulated into one service per domain. All of these individual services are pulled into the default target and, not having any dependencies blocking them, spring into action concurrently all at the same time. With a large number of domains and certs, this can overload the system. Resource constraints are not suitable for this, as they e.g. lead to timeouts of individual services that still start all at once, but now compete for even less compute time.

While in this particular case the individual services are not designed as template-instancing yet, they could be rebuilt this way.
Still I'd like to mention another implementation idea for this: Systemd could expose a kind of locking (semaphore) endpoint where units can attach to and block on. One option for this is implementing this on a slice, that only allows a certain maximum number of units to run within its scope, making all others block. This would align well with the existing resource constraint architecture, where limiting several different services is coordinated via slices as well.

**Comment by Werkov** [comment]
Just dropping induced brainstorm ideas.

### A motivation (dummy) example
```
cat >foo@.service <<EOD
[Unit]
After=foo@x%i.service # "follower"

[Install]
WantedBy=all-foos.target
EOD

for i in x xx ... xxxx ; do # follower enumeration
   systemctl enable foo@$i.service
done

systemctl start all-foos.target # ordering ensures concurrency of 1
```

### <del>Derived (semaphore) idea</del>

```
cat >foo@.service <<EOD
[Unit]
After=foos.semaphore
EOD
```

Job ordering wrt .semaphore unit would depend on its state (actually running instances vs configured capacity).

### Yet simpler idea

(I think you nailed it, @osnyx, with the semaphore idea -- [nothing extra](https://github.com/nathanweeks/ipcmd/tree/master) needs to be implemented.)
```
ExecStartPre=/usr/bin/sem_down $sem_name # blocking wait
ExecStart=...
ExecStartPre=/usr/bin/sem_up $sem_name
```

**Comment by osnyx** [comment]
@Werkov 

> # ordering ensures concurrency of 1

A concurrency of > 1 would be really useful, though. Could be achieved through "batching" of After dependencies, but that will get messy pretty quickly.

Regarding the ipcmd idea:
First of all thanks, this was the command I was failing to find all the time.

One thing we need to pay attention to is avoiding stale locks if the service fails, so how does your example behave if the `ExecStart` fails, will it unlock anyways?
Also, is 2x `ExecStartPre` correct?

In case we end up not implementing any locking/ limiting/ semaphore support into systemd, it still makes sense to properly document patterns like the usage of e.g. *ipcmd* somewhere. Otherwise we're back to the arcane knowledge of implicit complexity in sysvinit start scripts.

**Comment by Werkov** [comment]
(Sorry, I meant `ExecStartPost=`, fixed in the original comment.)
The directive's behavior is documented in systemd.service(5) (and also mentions `ExecStopPost=` that's even better for the cleanup).

Note, ipcmd was just one of the first search results for me, there may be more suitable implementation.

**Comment by osnyx** [comment]
@Werkov 
> ipcmd was just one of the first search results for me, there may be more suitable implementation.

Would you mind sharing your search terms? Because all the research I did in the end more or less converged on `flock` – which does not provide semaphore semantics – or on the linux kernel semaphore docs without mentioning userland CLI bindings for that.

**Comment by Werkov** [comment]
https://duckduckgo.com/?t=ffab&q=sysv+semaphores+cmdline&ia=web

**Comment by behrmann** [comment]
> There is a chance that https://github.com/systemd/systemd/pull/27985 unfortunately does not tackle this in its current state. Reading the OP's use case, we want all the started instances to succeed in the end, but just the ones exceeding a concurrency limit to block until there's room for them again.
The PR above so far reads like units exceeding such a limit will end up in a failed state instead of blocking and waiting, though.
> 
> Am I missing a way about how such a fail-on-limit behaviour could be turned into the units (re)starting until the limit is not exceeded anymore?

That's totally fine. The given PR doesn't solve this problem of batching instances/services, but failing additional instances/services is a legitimate failure mode (even better if it's a discernable error something like an orchestrator can react to). This issue is more about making the alternative available as well.

**Comment by behrmann** [comment]
@Werkov About the workarounds, those are indeed a possibility and some are already mentioned in the original post. I hacked up something similar once with an additional service that started units in a loop and waited with shell job control for some of them to finish before starting new ones. The point is, that that's cumbersome and has different edge cases, e.g. the `After=` example could easily have cases where the thing is too serialized and fewer than the allotted number of jobs is running (e.g. I have $k$ different lines of $x_{m+1}$ after $x_m$ that hold the $n$ instances and for one $k$ all jobs finish early, but this won't get backfilled from the other lines) or order becomes possibly unclear (e.g. because different jobs check the semaphore at slightly different times and race each other to start) and `ExecStartPre=` and `Post=` also shouldn't be used for long-running things and a semaphore might block indefinitely. 

So I think there's enough foot guns here to warrant a more userfriendly implementation where people don't need to reinvent the wheel. :)

**Comment by dtardon** [comment]
> There is a chance that #27985 unfortunately does **not** tackle this in its current state. Reading the OP's use case, we want all the started instances to succeed in the end, but just the ones exceeding a concurrency limit to block until there's room for them again.

systemd has no facility to postpone the start of a unit until a condition is fulfilled. In some cases this can be emulated (e.g., by adding artificial ordering dependencies), but there's no generic way to do it. I expect adding such a facility would be a _big_ and _invasive_ change; I'm not sure the gain--making some very specific workflows possible/easier--would be worth the additional complexity.

> The PR above so far reads like units exceeding such a limit will end up in a failed state instead of blocking and waiting, though.

Right, because the latter is not possible.

**Comment by osnyx** [comment]
@dtardon A comment of @poettering sounds like requeuing a failed unit – which is rather similar to blocking it until a condition is fulfilled – is either already possible or should not be hard to implement:

> Initially there'd be just two modes: "fail" and "delay". and the implementaiton would be easy: if "fail" is set we return a new error from unit_start() as you already implemented it. And "delay" would result in EAGAIN being returned, so that we simply retry later. A slice unit that uses "delay" mode should ensure though that all units with jobs contained in it are rechecked (via job_add_to_run_queue()) whenever a unit in the slice goes down.

https://github.com/systemd/systemd/pull/27985#issuecomment-1621702189

@Werkov I've played around a bit with both SysV and POSIX-Semaphores, unfortunately both have a fundamental flaw resulting in potentially stale locks:
If a process holding a semaphore (the process has decreased the semaphore at least once) gets killed, the lock is not released (semaphore not incremented) and blocked calls continue to block. So if enough of the processes crash, the remaining ones will stay deadlocked.

Or can we rely on `ExecStartPost`/ `ExecStopPost` being executed in **all** crash situations?

One advantage of getting this into systemd as a feature is that systemd itself could even use semaphores internally for that. Because if PID0 crashes you surely have more urgent problems than some services being blocked on a stale lock.

**Comment by sandy-lcq** [comment]
I met an similar issue,  I have a service template foo@service,  and runtimely,  foo@a.service foo@b.service ... will be enabled.  and the "instance name" is "a", "b" is randomly defined by user,  and the different instances don't have dependency.  After reboot,  
all the instances of foo@service started at the same time,  it make the whole system hang there since memory is not enough, even OOM.  so I also hope this getting this into systemd as a feature,  make service template can limit the number of instances,  if the limit number is reached,  other instances should be delayed.  or maybe other limit like cpu/memory,  if the used memory reached the limit,  eg:   80%,  then other instances are delayed.  And finally all the instances can start successfully.

**Comment by osnyx** [comment]
@sandy-lcq 
> or maybe other limit like cpu/memory, if the used memory reached the limit, eg: 80%, then other instances are delayed. And finally all the instances can start successfully.

Please note, that depending on the kind of service, resource limits are not a viable approach. Imposing CPU limits for example might slow down external network communications to an extent that they run into timeouts, like in this case: https://github.com/NixOS/nixpkgs/issues/232505#issuecomment-1587889374

In the end, I solved this particular problem with an elaborated locking wrapper to indeed limit concurrency of service startups. Integrating this into systemd would be very welcome indeed.

**Comment by poettering** [comment]
Lets mark this as duplicate of #35862, even though that one is newer. It's more generic, and covers both cases, and has a nice idea how to solve this.
